torch/csrc/jit/tensorexpr/function.cpp

```diff
@@ -99,44 +127,110 @@ Tensor* Compute(
-  std::vector<ExprHandle> strides(dims_.size());
-  for (size_t i = 0; i < strides.size(); i++) {
-    if (i == strides.size() - 1) {
-      strides[i] = ExprHandle(1);
-      continue;
-    }
-    ExprHandle stride = ExprHandle(dims_[i + 1]);
-    for (size_t j = i + 2; j < dims_.size(); j++) {
-      stride = stride * ExprHandle(dims_[j]);
-    }
-    strides[i] = stride;
+Tensor* ReduceCommon(
+    const std::string& func_name,
+    const Dtype& dtype,
+    std::vector<const Expr*>& outer_dims,
+    std::vector<const Var*>& outer_args,
+    ExprHandle initializer,
+    std::function<ExprHandle(const ExprHandle&)>& reduce_op,
+    std::vector<const Expr*>& reduce_dims,
+    std::vector<const Var*>& reduce_args) {
+  std::vector<const Expr*> all_dims;
+  all_dims.insert(all_dims.end(), outer_dims.begin(), outer_dims.end());
+  all_dims.insert(all_dims.end(), reduce_dims.begin(), reduce_dims.end());
+
+  std::vector<ExprHandle> strides;
+  getDimStrides(all_dims, strides);
+
+  const Var* func_var = new Var(func_name, kHandle);
+  const Expr* mask = new IntImm(1);
+  const Load* accum = new Load(
+      dtype,
+      func_var,
+      getUnrolledIndex(outer_dims, outer_args, strides).node(),
+      mask);
+
+  if (initializer.dtype() != dtype) {
+    initializer = Cast::make(dtype, initializer);
   }
 
-  ExprHandle total_index = int32_t{0};
-  for (size_t i = 0; i < dims_.size(); i++) {
-    ExprHandle index = VarHandle(this->args_[i]) * ExprHandle(strides[i]);
-    if (i == 0) {
-      total_index = index;
-    } else {
-      total_index = total_index + index;
-    }
-  }
+  const ExprHandle reduce_body = reduce_op(ExprHandle(accum));
+
+  Reduction* reduction = new Reduction(
+      func_name,
+      func_var,
+      outer_dims,
+      outer_args,
+      initializer.node(),
+      reduce_body.node(),
+      reduce_dims,
+      reduce_args);
+  return new Tensor(reduction, 0);
+}
 
+Stmt* Function::ElementStmt(size_t var_index, const Expr* body) {
+  std::vector<ExprHandle> strides;
+  getDimStrides(dims_, strides);
+  ExprHandle total_index = getUnrolledIndex(dims_, args_, strides);
   const Expr* mask = new IntImm(1);
 
-  Stmt* update_stmt =
-      new Store(func_var(index), total_index.node(), body(index), mask);
+  Store* update_stmt =
+      new Store(func_var(var_index), total_index.node(), body, mask);
   return update_stmt;
 }
 
+Stmt* Function::LoopStmt(Stmt* body) {
+  // TODO: verify that body is our body... how?
+  if (ndim() == 0) {
+    return body;
+  }
+
+  for (size_t i = 0; i < ndim(); i++) {
+    // Going in reverse order: from innermost loop to the outermost
+    size_t dim_index = ndim() - i - 1;
+    body = new For(arg(dim_index), new IntImm(0), dim(dim_index), body);
+  }
+  return body;
+}
+
+Stmt* Reduction::LoopStmt(Stmt* body) {
+  // TODO: Support multiple-output functions
+  // TODO: verify that body is our body... how?
+  if (ndim() + rdim() == 0) {
+    return body;
+  }
+
+  for (int i = 0; i < rdim(); ++i) {
```

Comment:
This structure dictates that the reduce axis are always the innermost. How would we reorder them in future?